test/table.js

```diff
@@ -1,128 +1,210 @@
+      const error = await catchError(table.read.bind(table))
+      assert.include(error.message, 'duplicates')
+    })
+
+    it('unique constraints violation for primary key', async () => {
+      const source = [
+        [1, '10.1', '1', 'string1', '2012-06-15'],
+        [1, '10.2', '2', 'string2', '2012-07-15'],
+      ]
+      const table = await Table.load(source, {schema: SCHEMA, headers: false})
+      const error = await catchError(table.read.bind(table))
+      assert.include(error.message, 'duplicates')
+    })
+
+    it('should read source data and limit rows', async () => {
+      const table = await Table.load(SOURCE, {schema: SCHEMA})
+      const rows = await table.read({limit: 1})
+      assert.deepEqual(rows.length, 1)
+    })
+
+    it('should read source data and return keyed rows', async () => {
+      const table = await Table.load(SOURCE, {schema: SCHEMA})
+      const rows = await table.read({keyed: true, limit: 1})
+      assert.deepEqual(rows[0],
+        {id: 1, height: 10.0, age: 1, name: 'string1', occupation: new Date(2012, 6-1, 15)})
+    })
+
+    it('should read source data and return extended rows', async () => {
+      const table = await Table.load(SOURCE, {schema: SCHEMA})
+      const rows = await table.read({extended: true, limit: 1})
+      assert.deepEqual(rows[0], [2,
+          ['id', 'height', 'age', 'name', 'occupation'],
+          [1, 10.0, 1, 'string1', new Date(2012, 6-1, 15)]])
+    })
+
+    it('should infer headers and schema', async function() {
+      if (process.env.USER_ENV === 'browser') this.skip()
+      const table = await Table.load('data/data_infer.csv')
+      await table.infer()
+      assert.deepEqual(table.headers, ['id', 'age', 'name'])
+      assert.deepEqual(table.schema.fields.length, 3)
+    })
+
+    it('should throw on read for headers/fieldNames missmatch', async () => {
+      const source = [
+        ['id', 'bad', 'age', 'name', 'occupation'],
+        [1, '10.0', 1, 'string1', '2012-06-15 00:00:00'],
+      ]
+      const table = await Table.load(source, {schema: SCHEMA})
+      const error = await catchError(table.read.bind(table))
+      assert.include(error.message, 'match schema field names')
+    })
 
-  it('should read source data and return keyed rows', async () => {
-    const table = await Table.load(SOURCE, {schema: SCHEMA})
-    const rows = await table.read({keyed: true, limit: 1})
-    assert.deepEqual(rows[0],
-      {id: 1, height: 10.0, age: 1, name: 'string1', occupation: new Date(2012, 6-1, 15)})
   })
 
-  it('should read source data and return extended rows', async () => {
-    const table = await Table.load(SOURCE, {schema: SCHEMA})
-    const rows = await table.read({extended: true, limit: 1})
-    assert.deepEqual(rows[0], [2,
-        ['id', 'height', 'age', 'name', 'occupation'],
-        [1, 10.0, 1, 'string1', new Date(2012, 6-1, 15)]])
-  })
+  describe('#foreignKeys', () => {
```

Comment:
Maybe add a test for self reference FK, as in the Python implementation? Can be done on a separate PR though.

Reply:
We have for self-referencing in `datapackage` tests.

Reply:
True, sorry I got the two mixed up!